### ISSUE TITLE
Bug: Self crawling rules don't respect unparsable code.

### ISSUE THREAD

**Issue by enote-kane** [sqlfluff/sqlfluff#3989]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

After upgrading from 1.2.1 to 1.3.0, previously linted files are showing `L003` errors.

### Expected Behaviour

Linting should not introduce regressions when upgrading sqlfluff within a major version (1.2.1 → 1.x).

Quite some tools (e.g. Flyway for database migrations) produce and verify file check sums, making it impossible to "fix" an older SQL file.

### Observed Behaviour

```bash
$ sqlfluff lint --nocolor --verbose --dialect postgres 2022-10-25.indentation-regression.sql
==== sqlfluff ====
sqlfluff:                1.3.0 python:                 3.8.10
implementation:        cpython verbosity:                   1
dialect:              postgres templater:               jinja
rules:                                all
==== readout ====

=== [ path: 2022-10-25.indentation-regression.sql ] ===

== [2022-10-25.indentation-regression.sql] FAIL                                                                                                                                   
L:   6 | P:   2 | L003 | Expected 1 indentation, found 0 [compared to line 05]
L:   8 | P:   1 | L003 | Expected 1 indentation, found 0 [compared to line 05]
==== summary ====
violations:        2 status:         FAIL
All Finished!
```

### How to reproduce

Reproduce with the following file:
```sql
CREATE OR REPLACE AGGREGATE range_agg_preserve_gaps(TSTZRANGE)  -- noqa
(
    sfunc = tstzrange_arr_merge,
    stype = TSTZRANGE[],
    initcond = '{}'
);

CREATE TABLE test (my_col TEXT);
```

Even when excluding the whole (unparseable) `AGGREGATE` block, the regression remains for the flawless `CREATE TABLE`:
```sql
-- noqa: disable=all
CREATE OR REPLACE AGGREGATE range_agg_preserve_gaps(TSTZRANGE)
(
    sfunc = tstzrange_arr_merge,
    stype = TSTZRANGE[],
    initcond = '{}'
);
-- noqa: enable=all

CREATE TABLE test (my_col TEXT);
```

Result:
```
== [2022-10-25.indentation-regression.sql] FAIL                                                                                                                                   
L:  10 | P:   1 | L003 | Expected 1 indentation, found 0 [compared to line 06]
==== summary ====
```

### Dialect

postgres

### Version

Regression starts with 1.3.0
```
sqlfluff:                1.3.0 python:                 3.8.10
implementation:        cpython verbosity:                   1
dialect:              postgres templater:               jinja
rules:                                all
```

...but continues through 1.3.2:
```
sqlfluff:                1.3.2 python:                 3.8.10
implementation:        cpython verbosity:                   1
dialect:              postgres templater:               jinja
rules:                                all
```

### Configuration

_none_

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
The second query is solved in #3942 which reworks L003.

I'll take a look at the first query to work out what's going on there.

**Comment by alanmcruickshank** [comment]
Ah yes - I think this is a regression based on the introduction of rule crawlers #3717.

The problem here is the interaction between unparsable sections and rule crawling. At the moment, unparsable code isn't yielded by the crawlers, but rules like L003 (which do their own crawling), don't know how to respect that.

I'll pick up this one. Will fix it before the next release.

@enote-kane - I'd suggest staying on pre-1.3.0 for now, indentation is going to change again with 2.0.0 - so that will mean you'll experience the least disruption on your side.

**Comment by alanmcruickshank** [comment]
Now that #4032 is merged, this is resolved with #3942. I've just tested our original query on that branch and no issues are raised.

**Comment by enote-kane** [comment]
@alanmcruickshank Thanx for the update, major version updates are expected to have breaking changes, so that's totally fine.
